### ISSUE TITLE
Can't assign union types to type arguments

### ISSUE THREAD

**Issue by falsandtru** [microsoft/TypeScript#36421]
I suppose the second error is an undesired effect of type checking to make the first error.

**TypeScript Version:**  3.7.x-dev.20200123

**Search Terms:**

**Code**

```ts
declare function concat<T>(target: T[], source: T[]): T[];
declare const as: number[];
declare const bs: string[];
concat(as, bs); // Correct error
declare const cs: number[] | string[];
concat(cs, []); // Incorrect error
concat(cs, cs); // Incorrect error
```

**Expected behavior:**
Error only with `concat(as, bs)`.
**Actual behavior:**
Error also with `concat(cs, [])` and `concat(cs, cs)`.
```
Argument of type 'number[] | string[]' is not assignable to parameter of type 'number[]'.
  Type 'string[]' is not assignable to type 'number[]'.(2345)
```
**Playground Link:** http://www.typescriptlang.org/play/index.html?target=99&ts=3.8.0-dev.20200123&ssl=1&ssc=1&pln=7&pc=35#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzDzCgwB4AVAPgAoNYBzEDALnnIG0BdAGngGccyGGBCsOnAJRiuAbgBQoSLASFUfDPCh9WqZAFsARiBiyF4aHAJ518A9v4YYWVPVOriGalt52JM+AD0AfAAwjgwcJjwxjDhZkqWqjZg9rqGxlzwAD4OTi5uRCTUKbxcfoHBAJLo4ZEaMXHuRSUEfOVB8NWEEeD1EeFAA

**Related Issues:**

**Comment by falsandtru** [comment]
This change unexpectedly makes `(number | string)[]`.
